### PR TITLE
SONARJAVA-3522 Rule S3414: Tests should be kept in a dedicated source directory

### PR DESCRIPTION
**Pull request by margarita-nedzelska-sonarsource**
The main idea is that main sources should not contain any classes with tests.

### CODE REVIEW

File: java-checks-test-sources/src/main/java/checks/NoTestsInProductionFile.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package checks;
+
+import org.junit.jupiter.api.Test;
+
+public class NoTestsInProductionFile { // Noncompliant [[sc=14;ec=35]] {{Move this test class to a separate assembly in this solution.}}
```
Comment (m-g-sonar) [inline java-checks-test-sources/src/main/java/checks/NoTestsInProductionFile.java:5]: Remove this `Noncompliant` message, and replace it by `compliant`, as we do not expect any issue here. I guess it's a copy-paste issue, but this is misleading.

File: java-checks-test-sources/src/main/java/checks/TestsInProductionFile.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package checks;
+
+import org.junit.jupiter.api.Test;
+
+public class TestsInProductionFile { // Noncompliant [[sc=14;ec=35]] {{Move this test class to a separate assembly in this solution.}}
```
Comment (m-g-sonar) [inline java-checks-test-sources/src/main/java/checks/TestsInProductionFile.java:5]: The issue message seems to come from C# world: `Move this test class to a separate assembly in this solution.`. I think we should have a different one for java, as this vocabulary does not apply to our context.

File: java-checks/src/main/java/org/sonar/java/checks/TestsInSeparateFolderCheck.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+    if (isTestClass(classTree)) {
+      reportIssue(classTree.simpleName(), "Move this test class to a separate assembly in this solution.");
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/TestsInSeparateFolderCheck.java:43]: The message does not seem appropriate for java projects.

**Review by m-g-sonar — CHANGES_REQUESTED** [review]

File: java-checks/src/main/java/org/sonar/java/checks/TestsInSeparateFolderCheck.java (hunk 3 above)
Comment (margarita-nedzelska-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/TestsInSeparateFolderCheck.java:43]: Maybe "Move this test class to the test folder in current project." will sound better?

**Review by m-g-sonar — APPROVED** [review]

**Pull request merged**